add pagination for keys in each lock

wawhal opened the pull request:
Fix #465 

![gif](https://media.giphy.com/media/2w5fvPom3YkkFJjMNN/giphy.gif)

## Proposed Changes

Adding pagination for the list of keys for each lock
- 10 items per page
- Can go to previous and next page
- Can jump to first page or last page
- Can jump to a page directly
- This pagination component can be reused for pagination purposes

Things that I had proposed but I could not see them fitting in the design provided in the issue:
- Giving ability to set the number of items per page

I am new to this crypto world and don't understand much of these contracts and stuff. I am a React developer and I worked on the pagination bit by using dummy data.

Let me know if there is anything you would like to change :-)

julien51 commented:
Thanks a lot @wawhal ! I'll review all that ASAP. Do you mind adding a screenshot (or even the screen recording you added on the issue, this will allow for a better idea of what is going on here). Thanks!

julien51 commented on unlock-app/src/__tests__/middlewares/currencyConversionMiddleware.test.js:
```diff
@@ -4,8 +4,12 @@ import { setConversionRate } from '../../actions/currencyconvert'
 afterEach(() => mockAxios.reset())
 
 describe('Currency conversion service retrieval middleware', () => {
```
Could you remove this file from your commit since these changes are not related to the purpose of your PR? Thanks!

julien51 replied:
This should be fixed once https://github.com/unlock-protocol/unlock/pull/732 has been merged. Could you please rebase after that? Thanks!

julien51 commented on unlock-app/src/components/creator/lock/KeyList.js:
```diff
@@ -3,26 +3,37 @@ import { connect } from 'react-redux'
+        <Row key={item.id}>
+          <Data>{item.transaction}</Data>
+          <Cell>{expirationAsDate(item.expiration)}</Cell>
+          <Data>{item.key}</Data>
```
that would be `item.data`. Also do you mind actually renaming the `item` variable into `key` as it is more explicit (or `lockKey` since `key` is also used by jsx)

julien51 commented on unlock-app/src/components/interface/pagination/Next.js:
```diff
@@ -0,0 +1,38 @@
+  }
+  return (
+    <ArrowGroup>
+      <RightArrow onClick={goToNextPage}>&gt;</RightArrow>
```
I think it would be good to maybe have a `disabled` state when there are no next page?

wawhal replied:
Not sure if that is possible. Because the button is a `div` and not a `button`. `div` doesn't take a `disabled` attribute.

julien51 replied:
I understand but you could make its "style" show that it is disabled (lighter?) and with no click possible on it. I tend to think that if it looks the same but behave differently, this is not a great experience... 
I actually think it would be better to use `<a>` rather than div in order to remain semantic here?

wawhal replied:
Makes sense. I'll make the changes.

julien51 commented on unlock-app/src/components/interface/pagination/PageNumbers.js:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  // If number of pages less than 10, show all 10 page numbers
+  if (numberOfPages <= 10) {
```
Maybe best to define `MAX_NUMBER_OF_PAGES_TO_SHOW_ALL`  constant in `constants.js`?

julien51 commented on unlock-app/src/components/interface/pagination/PageNumbers.js:
```diff
@@ -0,0 +1,93 @@
+  return (
+    <PageGroup>
+      {//show first five page numbers
+        Array(5)
```
Same here, please use a constant so it is really easy to change to another number should we need to!

julien51 commented on unlock-app/src/components/interface/pagination/Pagination.js:
```diff
@@ -0,0 +1,109 @@
+  constructor(props) {
+    super(props)
+    const itemCount = props.items.length,
+      itemsPerPage = 10
```
Another constant ;)

wawhal commented on unlock-app/src/__tests__/services/web3Service.test.js:
```diff
@@ -850,7 +850,10 @@ describe('Web3Service', () => {
       it('should emit a new transaction once it has been submitted', done => {
```
These files are automatically being changed as I commit my changes. Some reformatter is doing that. As you see, I have not made any change to them. Also, the package-lock is exactly the same as that in master, still it shows diff, I have no clue why :(

julien51 replied:
huh, this is strange. Ok thanks for trying anyway :(

wawhal commented:
@julien51 I have worked on the changes that you proposed and pushed the updates. Updated the PR with the GIF.
I have also added stories for the PageNumber component as it has quite a lot of edge cases.
From the UI testing point of view, I have added a test case for the PageNumber component.  
Rest of the components are fairly declarative and I haven't added tests for them. Let me know if you need tests for them anyway. :-)

julien51 commented on unlock-app/src/components/creator/lock/KeyList.js:
```diff
@@ -5,15 +5,14 @@ import UnlockPropTypes from '../../../propTypes'
-  const renderItems = items => {
-    return items.map(item => {
+  const renderItems = lockKeys => {
+    return lockKeys.map(lockKey => {
```
nit: destructure? `return lockKeys.map(({id, transaction, expiration, data})=> {` and then use them directly below?

wawhal commented:
@julien51 I resolved all the reviews that you requested. Except the package-lock :-(
I don't know what to do with it. I reset it to master and pushed it, it still shows conflict.
Please give me specific instructions on how to resolve it, or please resolve it in the UI before merging.
Sorry :-(

cellog commented:
@wawhal you can probably fix the package-lock.json problem by upgrading npm with

`npm i -G npm`

and then run npm i in the folder, it should revert package-lock.json back to original

wawhal commented:
The tests seem to be working locally for me. Not sure why the build is failing.

julien51 commented:
Thanks a lot for your work on this @wawhal . Please give a little more time to review the PR.

julien51 commented on unlock-app/src/__tests__/services/web3Service.test.js:
```diff
@@ -823,7 +823,10 @@ describe('Web3Service', () => {
       it('should emit a new transaction once it has been submitted', done => {
         expect.assertions(2)
         web3Service.sendTransaction = jest.fn((transaction, args, cb) => {
-          return cb(null, { event: 'transactionHash', args: { hash: '0x123' } })
```
Same here. This should be fixed once https://github.com/unlock-protocol/unlock/pull/732 has been merged. Could you please rebase after that? Thanks!

cellog commented on unlock-app/src/__tests__/components/interface/PaginationPageNumbers.js:
```diff
@@ -0,0 +1,30 @@
+import React from 'react'
+import * as rtl from 'react-testing-library'
+import 'jest-dom/extend-expect'
```
```suggested
```

This is loaded in the jest.setup.js

cellog commented on unlock-app/src/components/creator/CreatorLock.js:
```diff
@@ -78,7 +83,7 @@ export class CreatorLock extends React.Component {
         {!showEmbedCode &&
           showKeys && (
-          <LockPanel>
+          <LockPanel onClick={this.preventPropagation}>
```
Instead of using preventPropagation and binding it in the constructor, I recommend:

```js
<LockPanel onClick={e => e.preventPropagation}>
```

This is much simpler and achieves the same result

cellog commented on unlock-app/src/components/creator/lock/KeyList.js:
```diff
@@ -48,6 +58,10 @@ const mapStateToProps = (state, { lock }) => {
 
 export default connect(mapStateToProps)(KeyList)
 
+const KeyListWrapper = styled.div`
```
instead of creating a new styled.div, perhaps you can put this inside the Table definition? I think this is OK because we don't export it.

wawhal replied:
It does not feel semantically correct to do that. We can't mix up the table with pagination component. You might want to add more customizations to `Table` or `KeyListWrapper` in future as they serve different functionalities.

cellog commented on unlock-app/src/components/interface/pagination/PageNumbers.js:
```diff
@@ -0,0 +1,102 @@
+            return pageNumber(pn + 1)
+          })}
+      </PageGroup>
+    )
```
The code:

```js
Array(numberOfPages)
          .fill()
          .map((_, pn) => {
            return pageNumber(pn + 1)
          })
```

is used quite a bit in this component. Perhaps we can do this once, and use slices, so for example on line 41:

```js
{Allpages.slice(0, 5)}
```

and a similar usage on line 81 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice

cellog commented on unlock-app/src/components/interface/pagination/PageNumbers.js:
```diff
@@ -0,0 +1,102 @@
+            return pageNumber(pn + 1)
+          })}
+      {//show first few dots
+        currentPage - 1 === HALF_MAX_PAGES_TO_SHOW_ALL ? (
```
Could you extract this complex logic into some constants and then render them so it is easier to follow the render logic?

cellog commented on unlock-app/src/components/interface/pagination/Pagination.js:
```diff
@@ -0,0 +1,119 @@
+import { PGN_ITEMS_PER_PAGE } from '../../../constants'
+
+class Paginate extends React.Component {
+  constructor(props) {
```
Looks great! can you add propTypes?

```js
     static propTypes  = {
      ...
     }
```

You should `import UnlockPropTypes from 'src/propTypes'` and then use them. Add any you don't find.

Don't forget to run `npm run lint`! It will catch these kinds of omissions

cellog commented on unlock-app/src/stories/interface/Pagination.stories.js:
```diff
@@ -0,0 +1,73 @@
+import React from 'react'
```
fabulous!

julien51 commented:
@wawhal Any blocker that we can remove for you? Please let us know if you are stuck.

wawhal commented:
Hey guys!
I am on vacation right now. My bad, I should have informed earlier.
I'll be back on the 2nd Jan 2018. Please wait if you can or you can go ahead and modify the PR.
Apologies.

julien51 commented:
Thanks @wawhal! Happy new year! Please let us know if you need anything else!

wawhal commented:
Sorry for the delay folks. Will finish everything by 5th noon. Its been a while since this is open :D

wawhal commented:
Hello folks. I am unable to perform local development.
Every `build`, `dev`, `start` command is failing with an error of types `Cannot resolve ... in ....`
Can you tell me if anything has changed in local development? Some additional step that we need to perform? 
@julien51 @cellog

cellog commented:
Could you try running npm ci in root, smart-contracts, and unlock-app? That should reset your deps to pristine versions. At what stage do you see the error?

cellog commented:
If you don't get this resolved by Monday, I can pull your branch and try to sort it out then

wawhal commented:
@cellog I am unable to perform local development on master either.

This set of commands:

```
git clone git@github.com:unlock-protocol/unlock && cd unlock
npm ci
cd smart-contracts && npm ci
cd ../unlock-app && npm ci && npm run dev
```

fails with an error:

```
Saving artifacts...

> unlock-app@0.1.0 set-link /home/wawhal/oss/unlock/unlock-app
> run-script-os


> unlock-app@0.1.0 set-link:darwin:freebsd:linux:sunos /home/wawhal/oss/unlock/unlock-app
> pushd src && (test -e artifacts && rm -f artifacts) && ln -s ../../smart-contracts/build artifacts

sh: 1: pushd: not found
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! unlock-app@0.1.0 set-link:darwin:freebsd:linux:sunos: `pushd src && (test -e artifacts && rm -f artifacts) && ln -s ../../smart-contracts/build artifacts`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the unlock-app@0.1.0 set-link:darwin:freebsd:linux:sunos script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/wawhal/.npm/_logs/2019-01-05T17_41_28_377Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! unlock-app@0.1.0 set-link: `run-script-os`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the unlock-app@0.1.0 set-link script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/wawhal/.npm/_logs/2019-01-05T17_41_28_391Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! unlock-app@0.1.0 dev: `npm run setup-dev && npm run set-link && nodemon src/server.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the unlock-app@0.1.0 dev script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

cellog commented:
I see the issue. Can you run bash or zsh? sh does not have pushd. I am going to remove that line it has caused too many troubles

wawhal commented:
I am using bash @cellog
Anyways, I got it working temporarily by replacing `pushd` with `cd`.
Will push the changes in a bit.

cellog commented:
Ok, good to know, what os are you running? Will be good to add to our machine matrix once we get one up and running

wawhal commented:
I am on Ubuntu 16.04

cellog commented:
P s. If you feel like doing a 3 character PR to replace pushd with cd that would be awesome, just make sure we end up back in the right direction with a cd .. at the end of the line. If not, I'll do it

wawhal commented:
@cellog sure after this is merged and closed. The build is failing for some reasons. Everything works locally.

julien51 commented:
@wawhal I *think* you should be able to rebase and push again and the tests would run correctly this time. I am very very sorry about the issues you encountered. Please accept my apology.
We're very thankful for your contributions...

wawhal commented:
@julien51 hey no issues. These are common problems :)
There's a bug in one of my recent commits. Will push a fix in US time. Sorry for more delay.

julien51 commented:
Haaaa, I see that the netlify deploy *also* fails when missing dependencies. Let me add another PR. Sorry about that, again!

julien51 commented:
One more rebase please?

wawhal commented:
@julien51 are you sure it is the problem with the build system and not my PR? Locally everything seems to be working, the tests, storybook, UI etc. Is there anything else I should be adding/removing or checking as such?

julien51 commented:
cc @akeem It looks like the PR is failing due to tests in locksmith?

akeem commented:
Hi @wawhal , my apologies for this. The fault should now be resolved in the latest master, please pull in the latest changes

julien51 commented:
Thanks @akeem ! 
@wawhal I just did a test by submitting an external PR (#951) and tests passed.

Please rebase and push again. This time should be good!
Thanks and sorry again for the trouble.

julien51 commented:
@cellog Please do a final check for @wawhal ;)

cellog commented on unlock-app/src/__tests__/components/interface/PaginationPageNumbers.js:
```diff
@@ -0,0 +1,29 @@
+  return ({ children }) => children
+})
+
+afterEach(rtl.cleanup)
```
```suggestion
```

This is automatically done in our jest setup

cellog commented on unlock-app/src/stories/interface/PaginationPageNumbers.stories.js:
```diff
@@ -0,0 +1,73 @@
+    keyid: {
+      transaction: '0x23749328748932748932473298473289473298',
+      lock: '0xab7c74abc0c4d48d1bdad5dcb26153fc8780f83e',
+      expiration: Math.floor(new Date().getTime() / 1000) + 86400 * 30, // 30 days from right now
```
This needs to be changed, the snapshot tests will fail every new day :). You can see in the snapshots that "Feb 7, 2019" exists in the markup for every key, and that today, the snapshots failed and had to be updated to "Feb 8, 2019" Tomorrow they will need another update. Just hard-code a date here, it's fine for storybook purposes

cellog commented on unlock-app/src/stories/creator/KeyList.stories.js:
```diff
@@ -0,0 +1,91 @@
+          id: `${numberStr} ${j}`,
+          transaction: `0x23749328748932748932473298473289473298${i}${j}`,
+          lock: `0xab7c74abc0c4d48d1bdad5dcb26153fc8780f83e${numberStr}`,
+          expiration: Math.floor(new Date().getTime() / 1000) + 86400 * 30, // 30 days from right now
```
same comment, please hard-code a date here

cellog commented on unlock-app/src/components/interface/pagination/Pagination.js:
```diff
@@ -0,0 +1,130 @@
+import NextButtons from './Next'
+import PageNumbers from './PageNumbers'
+import { PGN_ITEMS_PER_PAGE } from '../../../constants'
+import UnlockPropTypes from '../../../propTypes'
```
```suggestion
import UnlockPropTypes from '../../../propTypes'
import PropTypes from 'prop-types'
```

See below

cellog commented on unlock-app/src/components/interface/pagination/Pagination.js:
```diff
@@ -0,0 +1,130 @@
+class Paginate extends React.Component {
+  static propTypes = {
+    items: UnlockPropTypes.items,
+    renderItems: UnlockPropTypes.method,
```
```suggestion
    renderItems: PropTypes.func,
```

cellog commented on unlock-app/src/propTypes.js:
```diff
@@ -66,6 +66,9 @@ export const element = PropTypes.oneOfType([PropTypes.func, PropTypes.element])
 export const delay = PropTypes.number
 
+export const items = PropTypes.array
+export const method = PropTypes.func
```
```suggestion
```

remove in favor of just using prop-types directly

cellog commented on unlock-app/src/propTypes.js:
```diff
@@ -75,11 +78,13 @@ export default {
   locks,
   key,
   keys,
+  method,
```
```suggestion
```

cellog commented on unlock-app/src/propTypes.js:
```diff
@@ -66,6 +66,9 @@ export const element = PropTypes.oneOfType([PropTypes.func, PropTypes.element])
 
 export const delay = PropTypes.number
 
+export const items = PropTypes.array
```
```suggestion
export const keyList = PropTypes.arrayOf(key)
```

wawhal replied:
This is not correct. If the pagination component is reused somewhere else in the future, it could be an array of something else.

cellog replied:
We will update it then, go ahead with this for now 👍, as we don't have plans for pagination of anything else yet.

cellog commented on unlock-app/src/propTypes.js:
```diff
@@ -75,11 +78,13 @@ export default {
   conversion,
   delay,
   element,
+  items,
```
```suggestion
  keyList,
```

cellog commented on unlock-app/src/components/interface/pagination/Pagination.js:
```diff
@@ -0,0 +1,130 @@
+
+class Paginate extends React.Component {
+  static propTypes = {
+    items: UnlockPropTypes.items,
```
```suggestion
    items: UnlockPropTypes.keyList,
```

cellog commented on unlock-app/src/stories/creator/KeyList.stories.js:
```diff
@@ -0,0 +1,91 @@
+      .forEach((_, j) => {
+        keys[`${numberStr} ${j}`] = {
+          id: `${numberStr} ${j}`,
+          transaction: `0x23749328748932748932473298473289473298${i}${j}`,
```
```suggestion
          transaction: `0x${i}${j}23749328748932748932473298473289473298`,
```

The end of the key is cut off so we can't actually see that the pagination works when clicking through pages

wawhal commented:
We should be good to go @cellog ?

julien51 commented on unlock-app/src/artifacts:
```diff
@@ -1 +1 @@
-../../smart-contracts/build/
\ No newline at end of file
+../../smart-contracts/build
```
I *think* we can remove from git, so sending a PR for this now.

hatgit commented:
Hi @julien51, and @cellog, following recent comments from the work by @wawhal, this is just a friendly reminder that if the work has been accepted, kindly login to your Gitcoin dashboard for this bounty where you can accept the work and pay out the bounty if you haven't already. Thanks!
